### DISCUSSION TITLE
Mirror directory structure in IDE virtual folders ('filters')

### DISCUSSION THREAD

**Discussion by BlackxSnow** [premake/premake-core#1845] [Q&A] [answered]
I'm using VS2019 and finding that the included files aren't being put into filters based on the directory structure as mentioned at the top of [this page](https://premake.github.io/docs/vpaths/). I'd like "proj/someModule/moduleMain.cpp" to be reflected by the IDE but I can't figure out how to do so (or why it's not doing so already as the default behaviour).

**Comment by KyrietS** [comment, marked as answer]
Let's say I have the following project structure:

```
MySolution/
├── MyModule
│   ├── include
│   │   └── someHeader.hpp
│   ├── src
│   │   └── main.cpp
│   └── vendor
│       └── someLibrary
│           └── someLibraryHeader.hpp
└── premake5.lua
```

Now I define my project in `premake5.lua` like this:

```lua
-- premake5.lua
workspace "MySolution"
    architecture "x86_64"

    configurations { 
        "Debug", 
        "Release"
    }
	
    project "MyModule"
        location "MyModule"
        kind "ConsoleApp"
        language "C++"
        cppdialect "C++17"

        files {
            "%{prj.name}/src/**.cpp",
            "%{prj.name}/include/**.hpp",
            "%{prj.name}/vendor/**.hpp"
        }

        includedirs {
            "%{prj.name}/include",
            "%{prj.name}/vendor"
        }
```

And when I generate a solution with `vs2019` my whole directory structure is mirrored by default:

![image](https://user-images.githubusercontent.com/19170699/160283359-9a2b7b3c-e778-4d06-b5bc-39f57eecbaab.png)

So what problem do you have exactly?
Can you show a simple example of what you tried to do?

**Reply by BlackxSnow** [reply]
It's now uh.. resolved itself. I'm not sure how, or if I was maybe somehow mistaken originally but it's correctly mirroring the file structure.

**Reply by nickclark2016** [reply]
My guess is that you were looking at the filters view instead of the file system view (visual studio has both).  If all you need is the file system view, VS does that out of the box without any need for Premake to do anything.

**Comment by furkancaglayan** [comment]
Looks like this only works if project location is same as source location. So it doesn't work when I do this:

`workspace "MySolution"
    architecture "x86_64"

    configurations { 
        "Debug", 
        "Release"
    }
	
    project "MyModule"
        location "%{wks.location}/%{prj.name}"
        kind "ConsoleApp"
        language "C++"
        cppdialect "C++17"

        files {
            "%{prj.name}/../../src/**.cpp",
        }

        includedirs {
        }`

Is this intended?

**Reply by KyrietS** [reply]
I have a following structure this time:
```
project
├── MySolution
│   ├── MyModule
│   │   ├── include
│   │   │   └── someHeader.hpp
│   │   └── vendor
│   │       └── someLibrary
│   │           └── someLibraryHeader.hpp
│   └── premake5.lua
└── src
    └── main.cpp
```
 And my configuration:

```lua
-- premake5.lua
workspace "MySolution"
    architecture "x86_64"

    configurations { 
        "Debug", 
        "Release"
    }
	
    project "MyModule"
        location "MyModule"
        kind "ConsoleApp"
        language "C++"
        cppdialect "C++17"

        files {
            "%{prj.name}/../../src/**.cpp",
            "%{prj.name}/include/**.hpp",
            "%{prj.name}/vendor/**.hpp"
        }

        includedirs {
            "%{prj.name}/include",
            "%{prj.name}/vendor"
        }
```

After running `premake5 vs2022` I got the following result:
![image](https://user-images.githubusercontent.com/19170699/230664182-555612fd-a5e0-45c7-9612-3e8b7802ccb5.png)

I can also see that the file `MyModule.vcxproj.filters` which is generated by Premake contains:

```xml
  <ItemGroup>
    <ClCompile Include="..\..\src\main.cpp">
      <Filter>src</Filter>
    </ClCompile>
  </ItemGroup>
```

It seems to be correct. 
What kind of structure/filter do you expect to see when you include files from outside the entire solution?

**Comment by furkancaglayan** [comment]
Hi, thanks for the answer. Filters are fine, but can I also retain the folder structure in src? 

![image](https://user-images.githubusercontent.com/21962498/230667883-f9d7ebe1-f4c0-4d4e-b703-8fab40d11356.png)

![image](https://user-images.githubusercontent.com/21962498/230667936-12d6a492-8018-4557-803a-67c338ce61f5.png)

**Reply by KyrietS** [reply]
Please use "Write a reply" under my latest answer instead of creating a new answer.

I have added 2 more folders:

```
project
├── MySolution
│   ├── MyModule
│   │   ├── include
│   │   │   └── someHeader.hpp
│   │   └── vendor
│   │       └── someLibrary
│   │           └── someLibraryHeader.hpp
│   └── premake5.lua
└── src
    ├── core
    │   └── myCore.cpp
    ├── editor
    │   └── editor.cpp
    └── main.cpp
```

I have made no changes to `premake5.lua`. I got the following result:

![image](https://user-images.githubusercontent.com/19170699/230673911-3e15ac05-25f4-4076-b035-ddaa23572bb9.png)

I use Premake 5.0.0-beta2. Also note that if you use [vpaths](https://premake.github.io/docs/vpaths/) then mirroring the filesystem may be overriden by your own config.
